File: Chevron/chevron-heur1.cpp
#include <cstdio>
#include <iostream>
#include <algorithm>
#include <map>
#include <vector>
#include <sstream>

using namespace std;

struct Facility{
	int id;
	double lat, lon;
	int maxOcc;
};

struct Worker{
	vector<string> worker_properties;

	Worker(vector<string> &w){
		worker_properties = w;
	}

	double lat, lon;
};

struct Equipment{
	string name;
	double p;
};


struct WorkOrder{
	static int ID;
	int id = ID++;
	int fac_id;
	string eqType;
	int eqId;
	int priority;
	int time;
};

int WorkOrder :: ID = 0;


map<int, Facility> facilities;
vector<Worker> workers;
map<int, WorkOrder> workOrders;

void init()
{
	freopen("facilities.txt","r",stdin);
	int id, maxOcc;
	double lat, lon;

	while(scanf("%d %lf %lf %d", &id, &lat, &lon, &maxOcc) != EOF)
	{
		facilities[id] = {id, lat, lon, maxOcc};
	}
	
	freopen("workers.txt", "r", stdin);
	string s;
	while (getline(cin, s)) 
	{
		string name = "";
		int i;
		size_t pos = 0;
		string token;
		string delimiter = " ";

		vector<string> worker_properties;

		while ((pos = s.find(delimiter)) != std::string::npos) {
		    token = s.substr(0, pos);
		    worker_properties.push_back(token);
		    s.erase(0, pos + delimiter.length());
		}

		worker_properties.push_back(s);

		workers.push_back(Worker(worker_properties));
	}

	freopen(stdin, "r", stdin);
}



void print()
{

	cout << "----------------FACILITIES--------------" << endl;
	for (map<int, Facility>::iterator it = facilities.begin(); it != facilities.end(); it++)
	{
	    cout << it->second.id << " " << it -> second.lat << " " << it->second.lon << " " << it->second.maxOcc<<endl;
	}
	cout << "----------------------------------------" << endl;
	cout << "----------------WORKERS-----------------" << endl;
	for (auto w : workers)
	{
		for (auto s : w.worker_properties)
		{
			cout << s << " ";
		}

		cout << endl;
	}

	cout << "----------------------------------------" << endl;

	cout << "----------------WORK_ORDERS--------------" << endl;
	for (map<int, WorkOrder>::iterator it = workOrders.begin(); it != workOrders.end(); it++)
	{
	    cout << it->second.id << " " << it -> second.fac_id << " " << it->second.eqType << " " << it->second.eqId<< " " << it->second.priority << " "<<it->second.time<<endl;
	}
	cout << "----------------------------------------" << endl;
}

void add_workorder(int fac_id, string eqType, int eqId, int priority, int time)
{
	WorkOrder wo;
	wo.fac_id = fac_id;
	wo.eqType = eqType;
	wo.eqId = eqId;
	wo.priority = priority;
	wo.time = time;

	workOrders[wo.id] = wo;
}

void add_workorder_wrapper()
{
	int fac_id, eqId, priority, time;
	string eqType;
	cin >> fac_id; cout << fac_id << endl;
	cin >> eqType; cout << eqType << endl;
	cin >> eqId; cout << eqId << endl;
	cin >> priority; cout << priority << endl;
	cin >> time; cout << time << endl;

	add_workorder(fac_id, eqType, eqId, priority, time);
}



int main()
{
	init();
	print();


	while(1)
	{
		int type;
		scanf("\n%d",&type);
		switch(type)
		{
			case 0 : add_workorder_wrapper();break;
			case 1 : add_workorder_wrapper();break;
			case 2 : add_workorder_wrapper();break;
			case 3 : print();break;
		}
	}
}

File: Chevron/chevron-heur1.py
from math import cos, asin, sqrt, pi


class Facility:
    def __init__(self, id, lat, lon, maxOcc):
        self.id = id
        self.lat = lat
        self.lon = lon
        self.maxOcc = maxOcc

    def __repr__(self) -> str:
        return str(self.__dict__)


class Worker:
    def __init__(self, name, properties):
        self.name = name
        self.properties = properties

    def __repr__(self) -> str:
        return str(self.__dict__)


class Equipment:
    def __init__(self, name, prob):
        self.name = name
        self.prob = prob

    def __repr__(self) -> str:
        return str(self.__dict__)


class WorkOrder:
    ID = 0

    def __init__(self, fac_id, eqType, eqId, priority, time):
        self.id = WorkOrder.ID
        WorkOrder.ID += 1
        self.fac_id = fac_id
        self.eqType = eqType
        self.eqId = eqId
        self.priority = priority
        self.time = time
        self.isActive = True

    def __repr__(self) -> str:
        return str(self.__dict__)


facilities = {}
workers = {}
workOrders = {}
equipment = {}


def init():
    f = open("facilities.txt", "r")
    for line in f:
        print(line)
        id = int(line.split("\t")[0])
        lat = float(line.split("\t")[1])
        lon = float(line.split("\t")[2])
        maxOcc = int(line.split("\t")[3])

        fac = Facility(id, lat, lon, maxOcc)

        facilities[id] = fac

    f = open("workers.txt", "r")
    for line in f:
        print(line)
        line = line.strip()
        name = line.split("\t")[0]
        properties = line.split("\t")[1].split(" ")
        worker = Worker(name, properties)
        workers[name] = worker

    f = open("equipment.txt", "r")
    for line in f:
        print(line)
        name = line.split("\t")[0]
        prob = float(line.split("\t")[1])
        equipment[name] = Equipment(name, prob)


def my_print():
    print(facilities)
    print(workers)
    print(equipment)
    print(workOrders)


def add_workorder(fac_id, eqType, eqId, priority, time):
    w = WorkOrder(fac_id, eqType, eqId, priority, time)
    workOrders[w.id] = w
    return w.id


def delete_workorder(id):
    try:
        workOrders.pop(id)
    except Exception:
        print("Key error workOrders")


def distance(lat1, lon1, lat2, lon2):
    p = pi / 180
    a = 0.5 - cos((lat2 - lat1) * p) / 2 + cos(lat1 * p) * cos(lat2 * p) * (1 - cos((lon2 - lon1) * p)) / 2
    return 12742 * asin(sqrt(a))


def heur1(id_worker, id_fac):
    sol_key = None
    max_dist = 100000000000

    for key, val in workOrders.items():
        if not val.isActive or val.eqType not in workers[id_worker].properties:
            continue
        if max_dist > distance(facilities[id_fac].lat, facilities[id_fac].lon, facilities[val.fac_id].lat,
                               facilities[val.fac_id].lon):
            max_dist = distance(facilities[id_fac].lat, facilities[id_fac].lon, facilities[val.fac_id].lat,
                                facilities[val.fac_id].lon)
            sol_key = key

    if sol_key == None:
        return None

    workOrders[sol_key].isActive = False
    return sol_key, workOrders[sol_key].fac_id


# returnign None if there is no workorder
def get_next_workorder(id_worker, id_fac):
    return heur1(id_worker, id_fac)


init()

add_workorder(1, "Pump", "P032", 5, 3)
add_workorder(3, "Conveyer", "Con391", 1, 9)
add_workorder(4, "Seperator", "Sep028", 2, 3)
add_workorder(5, "Sensor", "Sen826", 4, 1)
add_workorder(1, "Security", "Sec032", 1, 2)
add_workorder(5, "Electricity", "El087", 3, 2)
add_workorder(1, "Networking", "Net012", 3, 4)
my_print()

print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
print(get_next_workorder(workers['Bob'].name, 2))
